Initial Implementation of FTP Data task

saiHemak opened the pull request:
This PR includes the FTP data task implementation along with the FTPTestServer implementation .

saiHemak commented:
@pushkarnk @ianpartridge @alblue  Can you please review ?

alblue commented:
@swift-ci please test

xwu commented on Foundation/URLSession/URLSession.swift:
```diff
@@ -195,6 +195,7 @@ open class URLSession : NSObject {
     fileprivate static let registerProtocols: () = {
 	// TODO: We register all the native protocols here.
         let _ = URLProtocol.registerClass(_HTTPURLProtocol.self)
+	let _ = URLProtocol.registerClass(_FTPURLProtocol.self)
```
Why is this indentation like that? Also, you don't need to write `let _ =`; it's just `_ =`.

xwu commented on Foundation/URLSession/TransferState.swift:
```diff
@@ -100,6 +108,49 @@ extension _HTTPURLProtocol._TransferState {
+            return self
+        }
+        //FTP Status code 213 marks the end of the header and start of the transfer on data port
+        func  isCompleteHeader(_ headerLine: String) ->Bool {
```
Nit: please fix the spacing here; operators like `->` need one space on either side--I'm not certain that Swift currently complains about `->` specifically, but it certain does about other infix operators

While you're at it, please remove the extra space after `func` and after `return` on the following line.

xwu commented on TestFoundation/FTPServer.swift:
```diff
@@ -0,0 +1,276 @@
+    private var socketAddress1 = UnsafeMutablePointer<sockaddr_in>.allocate(capacity: 1)
+    private var connectionSocket: Int32!
+    var dataSocket: Int32! // data socket for communication
+    var dataSocketPort: UInt16! // data socket port ,should be sent as part of header
```
Nit: `, [space] ` not `[space] ,`.

xwu commented on TestFoundation/FTPServer.swift:
```diff
@@ -0,0 +1,276 @@
+            _ = try attempt("bind", valid: isZero, bind(dataSocket, addr, socklen_t(MemoryLayout<sockaddr>.size)))
+            var sockLen = socklen_t(MemoryLayout<sockaddr>.size)
+            _ = try attempt("listen", valid: isZero, listen(dataSocket, SOMAXCONN))
+            //open the data port asynchronously.Port should be opened before ESPV header communication.
```
Nit: please punctuate and capitalize comments consistently, with spaces after punctuation, etc.

xwu commented on TestFoundation/FTPServer.swift:
```diff
@@ -0,0 +1,276 @@
+            }
+        } else if(read.contains("STOR")) {
+            /*
+             Request is for Upload.As we are only dealing with data,Just read data and ignore
```
Nit: again, please punctuate and capitalize sentences consistently.

xwu commented on TestFoundation/TestFTP.swift:
```diff
@@ -0,0 +1,116 @@
+        ]
+    }
+    
+    let saveString =  "FTP Implementation to test FTP Upload,download and data tasks.Instead of sending file, we are sending the hardcoded data.We are going to test FTP data,Download and upload tasks with delegates & completion handlers.Creating the data here as we need to pass the count as part of the header.\r\n"
```
Nit: again, please punctuate and capitalize consistently.

xwu commented on Foundation/URLSession/TransferState.swift:
```diff
@@ -83,8 +83,16 @@ extension _HTTPURLProtocol._TransferState {
-    func byAppending(headerLine data: Data) throws -> _NativeProtocol._TransferState {
-        guard let h = parsedResponseHeader.byAppending(headerLine: data) else {
+    func byAppendingHTTP(headerLine data: Data) throws -> _NativeProtocol._TransferState {
+	 /// If the line is empty, it marks the end of the header, and the result
```
This should be a comment, not a doc comment. Please also fix the indentation for `- Note:`.

saiHemak commented:
@xwu  Thanks for the review ..Am addressing your comments

xwu commented on TestFoundation/TestFTP.swift:
```diff
@@ -0,0 +1,116 @@
+        ]
+    }
+    
+    let saveString =  "FTP implementation to test FTP upload,download and data tasks. Instead of sending a file, we are sending the hardcoded data.We are going to test FTP data,download and upload tasks with delegates & completion handlers. Creating the data here as we need to pass the count as part of the header.\r\n"
```
Nit: extra space after `=`; inside the string, make sure you have one space after each comma and period.

xwu commented on build.py:
```diff
@@ -445,6 +445,7 @@
 	'Foundation/URLSession/TransferState.swift',
 	'Foundation/URLSession/libcurl/libcurlHelpers.swift',
     'Foundation/URLSession/http/HTTPURLProtocol.swift',
+	'Foundation/URLSession/ftp/FTPURLProtocol.swift',
```
Can you fix the indenting in the line above?

xwu commented on build.py:
```diff
@@ -520,6 +521,7 @@
 foundation_tests = SwiftExecutable('TestFoundation', [
 	'TestFoundation/main.swift',
         'TestFoundation/HTTPServer.swift',
+	'TestFoundation/FTPServer.swift',
```
Can you fix the indenting among these lines?

xwu commented on TestFoundation/TestFTP.swift:
```diff
@@ -0,0 +1,116 @@
+   
+    func test_ftpdatatask() {
+        let ftpURL = "ftp://127.0.0.1:\(TestFTP.serverPort)/test.txt"
+        let  req = URLRequest(url: URL(string: ftpURL)!)
```
Nit: extra space after `let`.

xwu commented on TestFoundation/TestFTP.swift:
```diff
@@ -0,0 +1,116 @@
+    
+    static var allTests: [(String, (TestFTP) -> () throws -> Void)] {
+        return [
+            
```
Please remove this extraneous empty line.

xwu commented on TestFoundation/FTPServer.swift:
```diff
@@ -0,0 +1,274 @@
+            try respondWithRawData(with: "150 Opening BINARY mode data connection for test.txt (\(dataCount!) bytes).\r\n")
+            // Send data here through data port
+            do {
+                let  dataWritten = try respondWithData(with: saveData!)
```
Nit: extra space after `let`.

xwu commented on TestFoundation/FTPServer.swift:
```diff
@@ -0,0 +1,274 @@
+    }
+    
+    public func respondWithData(with data: Data) throws -> Int32 {
+        return  try self.socket.writeRawData(socket: data)
```
Nit: extra space after `return`.

xwu commented on TestFoundation/FTPServer.swift:
```diff
@@ -0,0 +1,274 @@
+    override class func setUp() {
+        super.setUp()
+        func runServer(with condition: ServerSemaphore, startDelay: TimeInterval? = nil, sendDelay: TimeInterval? = nil, bodyChunks: Int? = nil) throws {
+            let start =   21961 //21 21961
```
Nit: extra spaces after `=`.

xwu commented on TestFoundation/FTPServer.swift:
```diff
@@ -0,0 +1,274 @@
+        _ = try data.withUnsafeBytes { ptr in
+            try attempt("write", valid: isNotNegative, CInt(write(connectionSocket, ptr, data.count)))
+        }
+        
```
Nit: extraneous empty line.

xwu commented on TestFoundation/FTPServer.swift:
```diff
@@ -0,0 +1,274 @@
+        
+    }
+ 
+    func writeRawData(socket data: Data) throws -> Int32{
```
Nit: needs space before `{`.

xwu commented on Foundation/URLSession/Message.swift:
```diff
@@ -49,24 +49,24 @@ extension _NativeProtocol._ParsedResponseHeader {
         let lineBuffer = data.subdata(in: Range(data.startIndex..<data.endIndex-2))
         guard let line = String(data: lineBuffer, encoding: String.Encoding.utf8) else { return nil}
-        return byAppending(headerLine: line)
+        return byAppending(headerLine: line, headerCompleted: headerCompleted)
```
I'm surprised and delighted that this happens to call the correct (private) overload, but given that the two methods have the same name and differ only by optionality of the return type, these is very confusing to read.

Please rename the private method using a leading underscore (that is, call it `_byAppending`) and call that here. It would be much clearer to the reader.

xwu commented on TestFoundation/FTPServer.swift:
```diff
@@ -0,0 +1,274 @@
+                    self.dataSocket = try self.attempt("accept", valid: self.isNotNegative, accept(self.dataSocket, addr, &sockLen))
+                    self.dataSocketPort = sa1.sin_port
+                } catch {
+                }
```
I know this is test code, but do you really mean to eat any error and continue? Seems like the test should abort.

xwu replied:
Isn't the entire purpose of the test suite to log failures?

alblue commented:
@swift-ci please test

alblue commented:
@swift-ci please test

saiHemak commented:
@xwu @alblue  Can you please review and let me know if anything is required ?

alblue commented:
@swift-ci please test

xwu commented:
There are some pervasive issues with code indentation, including those I pointed out above and new ones. Could you please run your changes through a linter?

ianpartridge commented:
@swift-ci please test

ianpartridge commented on Foundation/URLSession/Message.swift:
```diff
@@ -49,24 +49,24 @@ extension _NativeProtocol._ParsedResponseHeader {
     /// - Returns: Returning nil indicates failure. Otherwise returns a new
     ///     `ParsedResponseHeader` with the given line added.
-    func byAppending(headerLine data: Data) -> _NativeProtocol._ParsedResponseHeader? {
+    func byAppending(headerLine data: Data, headerCompleted: (String) -> Bool) -> _NativeProtocol._ParsedResponseHeader? {
```
Please document the parameters in the comments above the function.

ianpartridge replied:
I think the new parameter would be better named `onHeaderCompleted` to indicate that it's a callback and not a simple boolean.

ianpartridge replied:
It's not a boolean value though - it's a closure?

saiHemak replied:
Yeah I meant the closure which does the check

ianpartridge replied:
The parameter name refers to the whole parameter though, so it should make it clear that the parameter is a closure, not a boolean. `headerCompleted` sounds like it's a boolean that is being passed in ("is the header already completed?")

ianpartridge commented on Foundation/URLSession/TransferState.swift:
```diff
@@ -100,6 +108,49 @@ extension _HTTPURLProtocol._TransferState {
+// specific to FTP
+extension _FTPURLProtocol._TransferState {
+
+    enum FTPHeaderCode: Int {
```
Is there a reason these are out of order?

ianpartridge commented on Foundation/URLSession/TransferState.swift:
```diff
@@ -100,6 +108,49 @@ extension _HTTPURLProtocol._TransferState {
+    func byAppendingFTP(headerLine data: Data, expectedContentLength: Int64) throws -> _NativeProtocol._TransferState {
+        let line = String(data: data, encoding: String.Encoding.utf8)
+        //FTP Status code 226 marks the end of the transfer
+        if (line?.starts(with: String(FTPHeaderCode.transferCompleted.rawValue)))! {
```
Force unwraps are not good, please do appropriate safe error handling.

ianpartridge commented on Foundation/URLSession/TransferState.swift:
```diff
@@ -100,6 +108,49 @@ extension _HTTPURLProtocol._TransferState {
+
+        if case .complete(let lines) = h {
+            let response = lines.createURLResponse(for: url, contentLength: expectedContentLength)
+            guard response != nil else {
```
This can just be `guard response else`

ianpartridge commented on Foundation/URLSession/URLSession.swift:
```diff
@@ -194,7 +194,8 @@ open class URLSession : NSObject {
     fileprivate static let registerProtocols: () = {
 	// TODO: We register all the native protocols here.
-        let _ = URLProtocol.registerClass(_HTTPURLProtocol.self)
+       _ = URLProtocol.registerClass(_HTTPURLProtocol.self)
```
Should we remove the TODO here?

saiHemak replied:
I presume No .We have only implementation of HTTP and FTP in place as of now .Still we may need to extend the protocol support to FILE or DATA  or any other native protocols .Hence having this will help implementers to know  where to register the same ..

ianpartridge commented on TestFoundation/FTPServer.swift:
```diff
@@ -0,0 +1,275 @@
+    private var connectionSocket: Int32!
+    var dataSocket: Int32! // data socket for communication
+    var dataSocketPort: UInt16! // data socket port, should be sent as part of header
+    private func isNotNegative(r: CInt) -> Bool {
```
Bad name, based on the code it should be `isNotMinusOne()`

saiHemak replied:
Actually this method implementation is same as the ```isNotNegative``` method in HTTPServer .. Do you want me to rename that also ?

saiHemak commented:
@ianpartridge  I have addressed your comments except for two .Please review

ianpartridge commented:
Thanks for addressing my feedback, I see a lot of @xwu's is still valid though.

ianpartridge commented:
@swift-ci please test

ianpartridge commented:
@swift-ci please test

millenomi commented:
@saiHemak @ianpartridge is this PR abandoned?

saiHemak commented:
@millenomi  I have addressed all the comments ..I am not sure why it has not been merged .  If its good to go then I will rebase the branch to resolve the conflicts .

millenomi commented:
Thank you!

andrewtheis commented:
@saiHemak Looks like there are still some outstanding formatting issues that haven’t been addressed.

ianpartridge commented:
I'm afraid @saiHemak is currently on maternity leave from IBM so won't be able to work on this in the near future.

saiHemak commented:
@andrewtheis  Thanks for the review ..I  am working on it

pushkarnk commented:
@swift-ci please test

pushkarnk commented:
@swift-ci please test

pushkarnk commented:
@saiHemak CI failures, can you please take a look?

pushkarnk commented:
@swift-ci test

saiHemak commented:
@pushkarnk  I fixed the build issues and updated CMakeFile changes ..Can you please trigger build ?

spevans commented:
@swift-ci test

pushkarnk commented:
@swift-ci test

pushkarnk commented:
@swift-ci test

pushkarnk commented:
@swift-ci test

spevans commented:
@swift-ci test

pushkarnk commented:
@swift-ci test

pushkarnk commented:
@swift-ci test

pushkarnk commented:
@swift-ci test Linux

pushkarnk commented:
@swift-ci test Linux

pushkarnk commented:
@swift-ci test Linux

pushkarnk commented:
@swift-ci test

pushkarnk commented:
@swift-ci test

pushkarnk commented:
@swift-ci please test

pushkarnk commented:
@swift-ci please test

pushkarnk commented:
@swift-ci test

saiHemak commented:
@pushkarnk   Build failure seems to be irrelevant  ..Build failed due to unable to find the module ```CoreFoundation```. Can you please re-trigger on Linux

spevans commented:
@swift-ci test linux

pushkarnk commented:
@swift-ci test linux

pushkarnk commented:
@shahmishal @spevans The PR build seems to be failing with different apparently unrelated problems. Could you help us with a clean build please?

shahmishal commented:
@eeckstein is this known failure? 

```
23:11:32 ******************** TEST 'Swift(linux-x86_64) :: Prototypes/PatternMatching.swift' FAILED ********************
23:11:32 Script:
23:11:32 --
23:11:32 : 'RUN: at line 12';   rm -rf "/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/test-linux-x86_64/Prototypes/Output/PatternMatching.swift.tmp" && mkdir -p "/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/test-linux-x86_64/Prototypes/Output/PatternMatching.swift.tmp" && /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swiftc -target x86_64-unknown-linux-gnu  -module-cache-path '/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/swift-test-results/x86_64-unknown-linux-gnu/clang-module-cache' -swift-version 4  -O -module-cache-path '/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/swift-test-results/x86_64-unknown-linux-gnu/clang-module-cache' /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift -o /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/test-linux-x86_64/Prototypes/Output/PatternMatching.swift.tmp/a.out -module-name main  && echo /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/test-linux-x86_64/Prototypes/Output/PatternMatching.swift.tmp/a.out &&/usr/bin/env DYLD_LIBRARY_PATH='/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/lib/swift/linux/x86_64' LD_LIBRARY_PATH='/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/lib/swift/linux/x86_64' SIMCTL_CHILD_DYLD_LIBRARY_PATH='/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/lib/swift/linux/x86_64'  /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/test-linux-x86_64/Prototypes/Output/PatternMatching.swift.tmp/a.out
23:11:32 --
23:11:32 Exit Code: 254
23:11:32 
23:11:32 Command Output (stderr):
23:11:32 --
23:11:32 /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift:39:19: warning: redundant conformance constraint 'C.SubSequence': 'Collection'
23:11:32   , C.SubSequence : Collection  
23:11:32                   ^
23:11:32 /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift:36:19: note: conformance constraint 'C.SubSequence': 'Collection' implied here
23:11:32   func matched<C: Collection>(atStartOf c: C) -> MatchResult<Index, MatchData>
23:11:32                   ^
23:11:32 /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift:46:19: warning: redundant conformance constraint 'C.SubSequence': 'Collection'
23:11:32   , C.SubSequence : Collection  
23:11:32                   ^
23:11:32 /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift:43:17: note: conformance constraint 'C.SubSequence': 'Collection' implied here
23:11:32   func found<C: Collection>(in c: C) -> (extent: Range<Index>, data: MatchData)?
23:11:32                 ^
23:11:32 /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift:76:19: warning: redundant conformance constraint 'C.SubSequence': 'Collection'
23:11:32   , C.SubSequence : Collection  
23:11:32                   ^
23:11:32 /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift:73:19: note: conformance constraint 'C.SubSequence': 'Collection' implied here
23:11:32   func matched<C: Collection>(atStartOf c: C) -> MatchResult<Index, ()>
23:11:32                   ^
23:11:32 /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift:97:19: warning: redundant conformance constraint 'C.SubSequence': 'Collection'
23:11:32   , C.SubSequence : Collection  
23:11:32                   ^
23:11:32 /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift:94:19: note: conformance constraint 'C.SubSequence': 'Collection' implied here
23:11:32   func matched<C: Collection>(atStartOf c: C) -> MatchResult<Index, ()>
23:11:32                   ^
23:11:32 /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift:130:19: warning: redundant conformance constraint 'C.SubSequence': 'Collection'
23:11:32   , C.SubSequence : Collection  
23:11:32                   ^
23:11:32 /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift:127:19: note: conformance constraint 'C.SubSequence': 'Collection' implied here
23:11:32   func matched<C: Collection>(atStartOf c: C) -> MatchResult<Index, MatchData>
23:11:32                   ^
23:11:32 /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift:170:19: warning: redundant conformance constraint 'C.SubSequence': 'Collection'
23:11:32   , C.SubSequence : Collection  
23:11:32                   ^
23:11:32 /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift:167:19: note: conformance constraint 'C.SubSequence': 'Collection' implied here
23:11:32   func matched<C: Collection>(atStartOf c: C) -> MatchResult<M0.Index, MatchData>
23:11:32                   ^
23:11:32 /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift:240:19: warning: redundant conformance constraint 'C.SubSequence': 'Collection'
23:11:32   , C.SubSequence : Collection  
23:11:32                   ^
23:11:32 /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift:237:19: note: conformance constraint 'C.SubSequence': 'Collection' implied here
23:11:32   func matched<C: Collection>(atStartOf c: C) -> MatchResult<Index, MatchData>
23:11:32                   ^
23:11:32 /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift:296:19: warning: redundant conformance constraint 'C.SubSequence': 'Collection'
23:11:32   , C.SubSequence : Collection  
23:11:32                   ^
23:11:32 /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift:293:19: note: conformance constraint 'C.SubSequence': 'Collection' implied here
23:11:32   func matched<C: Collection>(atStartOf c: C) -> MatchResult<Index, ()>
23:11:32                   ^
23:11:32 /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift:331:19: warning: redundant conformance constraint 'C.SubSequence': 'Collection'
23:11:32   , C.SubSequence : Collection {
23:11:32                   ^
23:11:32 /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift:326:22: note: conformance constraint 'C.SubSequence': 'Collection' implied here
23:11:32   func searchTest<C: Collection>(
23:11:32                      ^
23:11:32 /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift:386:19: warning: redundant conformance constraint 'C.SubSequence': 'Collection'
23:11:32   , C.SubSequence : Collection  
23:11:32                   ^
23:11:32 /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift:383:19: note: conformance constraint 'C.SubSequence': 'Collection' implied here
23:11:32   func matched<C: Collection>(atStartOf c: C) -> MatchResult<Index, MatchData>
23:11:32                   ^
23:11:32 Stack dump:
23:11:32 0.	Program arguments: /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift -frontend -c -primary-file /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift/test/Prototypes/PatternMatching.swift -target x86_64-unknown-linux-gnu -disable-objc-interop -module-cache-path /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/swift-test-results/x86_64-unknown-linux-gnu/clang-module-cache -swift-version 4 -O -module-name main -o /tmp/lit_tmp_KtIUbP/PatternMatching-a9ee64.o 
23:11:32 1.	Swift version 5.0-dev (LLVM 77814f897e, Swift 6ca706eb1d)
23:11:32 2.	While running pass #42621 SILFunctionTransform "DeadStoreElimination" on SILFunction "@$sSi8midPoint_yt_ytt4datatSSIegno_SiSSIegyo_TR185$s4main7PatternPAAs5UInt8V7ElementRtzrlE10searchTest2in6formatyqd___SS9MatchDataQzXEtSlRd__5IndexQyd__AMRtzAeFRtd__lFfA0_SSALcfU_AA18ConsecutiveMatchesVyAA0I12StaticStringVATG_SayAEGTG5Tf3npf_n".
23:11:32  for <<debugloc at "<compiler-generated>":0:0>> #0 0x00000000048e5854 PrintStackTraceSignalHandler(void*) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0x48e5854)
23:11:32  #1 0x00000000048e3470 llvm::sys::RunSignalHandlers() (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0x48e3470)
23:11:32  #2 0x00000000048e5c78 SignalHandler(int) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0x48e5c78)
23:11:32  #3 0x00007f5e506e0390 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x11390)
23:11:32  #4 0x0000000000da52e6 swift::SILFunction::getTypeLowering(swift::SILType) const (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0xda52e6)
23:11:32  #5 0x0000000000e10b1f swift::SILType::isAddressOnly(swift::SILFunction const&) const (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0xe10b1f)
23:11:32  #6 0x0000000000b8e730 mayContainReference(swift::SILType, swift::SILFunction const&) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0xb8e730)
23:11:32  #7 0x0000000000b89219 swift::EscapeAnalysis::isPointer(swift::ValueBase*) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0xb89219)
23:11:32  #8 0x0000000000b88d73 swift::EscapeAnalysis::ConnectionGraph::getNode(swift::ValueBase*, swift::EscapeAnalysis*, bool) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0xb88d73)
23:11:32  #9 0x0000000000b9050f swift::EscapeAnalysis::analyzeInstruction(swift::SILInstruction*, swift::EscapeAnalysis::FunctionInfo*, swift::BottomUpIPAnalysis::BottomUpFunctionOrder<swift::EscapeAnalysis::FunctionInfo>&, int) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0xb9050f)
23:11:32 #10 0x0000000000b8ef94 swift::EscapeAnalysis::buildConnectionGraph(swift::EscapeAnalysis::FunctionInfo*, swift::BottomUpIPAnalysis::BottomUpFunctionOrder<swift::EscapeAnalysis::FunctionInfo>&, int) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0xb8ef94)
23:11:32 #11 0x0000000000b92071 swift::EscapeAnalysis::recompute(swift::EscapeAnalysis::FunctionInfo*) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0xb92071)
23:11:32 #12 0x0000000000b9339d swift::EscapeAnalysis::canEscapeTo(swift::SILValue, swift::FullApplySite) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0xb9339d)
23:11:32 #13 0x0000000000baebfa swift::SILInstructionVisitor<(anonymous namespace)::MemoryBehaviorVisitor, swift::SILInstruction::MemoryBehavior>::visit(swift::SILInstruction*) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0xbaebfa)
23:11:32 #14 0x0000000000bac75b swift::AliasAnalysis::computeMemoryBehaviorInner(swift::SILInstruction*, swift::SILValue, swift::RetainObserveKind) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0xbac75b)
23:11:32 #15 0x0000000000bac46a swift::AliasAnalysis::computeMemoryBehavior(swift::SILInstruction*, swift::SILValue, swift::RetainObserveKind) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0xbac46a)
23:11:32 #16 0x0000000000a36f6c (anonymous namespace)::DSEContext::processUnknownReadInst(swift::SILInstruction*, DSEKind) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0xa36f6c)
23:11:32 #17 0x0000000000a34c43 (anonymous namespace)::DSEContext::processInstruction(swift::SILInstruction*, DSEKind) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0xa34c43)
23:11:32 #18 0x0000000000a313f8 (anonymous namespace)::DeadStoreElimination::run() (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0xa313f8)
23:11:32 #19 0x0000000000993733 swift::SILPassManager::runPassOnFunction(unsigned int, swift::SILFunction*) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0x993733)
23:11:32 #20 0x0000000000994401 swift::SILPassManager::runFunctionPasses(unsigned int, unsigned int) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0x994401)
23:11:32 #21 0x00000000009955be swift::SILPassManager::execute() (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0x9955be)
23:11:32 #22 0x00000000005891e8 swift::SILPassManager::executePassPipelinePlan(swift::SILPassPipelinePlan const&) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0x5891e8)
23:11:32 #23 0x000000000099d0b0 swift::runSILOptimizationPasses(swift::SILModule&) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0x99d0b0)
23:11:32 #24 0x00000000007968e1 swift::CompilerInstance::performSILProcessing(swift::SILModule*, swift::UnifiedStatsReporter*) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0x7968e1)
23:11:32 #25 0x0000000000508c95 performCompile(swift::CompilerInstance&, swift::CompilerInvocation&, llvm::ArrayRef<char const*>, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0x508c95)
23:11:32 #26 0x0000000000504b9f swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0x504b9f)
23:11:32 #27 0x00000000004a002d main (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0x4a002d)
23:11:32 #28 0x00007f5e4ee0a830 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x20830)
23:11:32 #29 0x000000000049fb89 _start (/home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/buildbot_linux/swift-linux-x86_64/bin/swift+0x49fb89)
23:11:32 
<unknown>:0: error: unable to execute command: Segmentation fault
23:11:32 <unknown>:0: error: compile command failed due to signal 11 (use -v to see invocation)
23:11:32 
23:11:32 --
23:11:32 
23:11:32 ********************
```

pushkarnk commented:
@swift-ci please test Linux

saiHemak commented:
@pushkarnk  can you please review

pushkarnk commented on Foundation/URLSession/ftp/FTPURLProtocol.swift:
```diff
@@ -0,0 +1,117 @@
+
+    override class func canInit(with request: URLRequest) -> Bool {
+        guard request.url?.scheme == "ftp"
+            /* || request.url?.scheme == "ftps" || request.url?.scheme == "sftp" */
```
@saiHemak Can we remove this commented code and add a TODO here?

pushkarnk commented:
@saiHemak Would you mind renaming `TestFTP.swift` to something like `TestFTPURLSession.swift`. IMO the name must indicate that it's relevance to `URLSession`.

pushkarnk commented:
@swift-ci please test

pushkarnk commented:
@swift-ci please test

pushkarnk commented:
@swift-ci test and merge

pushkarnk commented:
@swift-ci test and merge

compnerd commented:
This change breaks the WIndows build
